### PR TITLE
APR-1278 - Create endpoint to Snapshot an Application Data and Files

### PR DESCRIPTION
**Pull request by martin-hornsby**

### CODE REVIEW

File: src/SFA.DAS.QnA.Application.UnitTests/CommandsTests/CreateSnapshotTests/CreateSnapshotTestBase.cs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+using SFA.DAS.QnA.Application.Commands.CreateSnapshot;
+using Microsoft.Extensions.Logging;
+using Microsoft.Azure.Storage.Blob;
+using SFA.DAS.QnA.Application.Commands.Files;
+using System.Text;
+using System.IO;
+
+namespace SFA.DAS.QnA.Application.UnitTests.CommandsTests.CreateSnapshotTests
+{
+    [TestFixture]
```
Comment (David-B-Read) [inline src/SFA.DAS.QnA.Application.UnitTests/CommandsTests/CreateSnapshotTests/CreateSnapshotTestBase.cs]: I think this needs to move into an integration tests project, it won't be able to run on Azure Devops

File: src/SFA.DAS.QnA.Application/Commands/CreateSnapshot/CreateSnapshotHandler.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+            await _dataContext.SaveChangesAsync(cancellationToken);
+            _logger.LogInformation($"Created ApplicationSection entities for Application: {newApplication.Id}");
+        }
+
+        private async Task CopyFiles(Data.Entities.Application currentApplication, Data.Entities.Application newApplication, CancellationToken cancellationToken)
+        {
+            var sections = await _dataContext.ApplicationSections.AsNoTracking().Where(sec => sec.ApplicationId == currentApplication.Id).ToListAsync();
+            var newSections = await _dataContext.ApplicationSections.AsNoTracking().Where(sec => sec.ApplicationId == newApplication.Id).ToListAsync();
+
+            // go through each section
```
Comment (MarkFCain) [inline src/SFA.DAS.QnA.Application/Commands/CreateSnapshot/CreateSnapshotHandler.cs]: no need for this comment

File: src/SFA.DAS.QnA.Application/Commands/CreateSnapshot/CreateSnapshotHandler.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            foreach (var section in sections)
+            {
+                var newSection = newSections.FirstOrDefault(s => s.SectionNo == section.SectionNo && s.SequenceNo == section.SequenceNo);
+                if (newSection is null) continue;
+
+                // go through each page within the section
```
Comment (MarkFCain) [inline src/SFA.DAS.QnA.Application/Commands/CreateSnapshot/CreateSnapshotHandler.cs]: no need for this comment

File: src/SFA.DAS.QnA.Application/Commands/CreateSnapshot/CreateSnapshotHandler.cs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+                                        // it exists so copy it into the new application
+                                        if (currentFileBlobReference.Exists())
+                                        {
+                                            var snapshotFileBlobReference = container.GetBlockBlobReference(snapshotFileUrl);
+                                            await snapshotFileBlobReference.StartCopyAsync(currentFileBlobReference);
+                                        }
+                                    }
+                                    catch (Exception ex)
+                                    {
+                                        _logger.LogError($"Error coping file in snapshot: {snapshotFileUrl} || Message: {ex.Message} || Stack trace: {ex.StackTrace}");
```
Comment (David-B-Read) [inline src/SFA.DAS.QnA.Application/Commands/CreateSnapshot/CreateSnapshotHandler.cs]: Typo in log message

File: src/SFA.DAS.QnA.Application.UnitTests/CommandsTests/CreateSnapshotTests/CreateSnapshotTestBase.cs (hunk 1 above)
Comment (martin-hornsby) [reply src/SFA.DAS.QnA.Application.UnitTests/CommandsTests/CreateSnapshotTests/CreateSnapshotTestBase.cs]: Fixed it for now using PreProcessor directives. Another option would be to mock it

File: src/SFA.DAS.QnA.Api/Controllers/ApplicationSnapshotController.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        private readonly IMediator _mediator;
+
+        public ApplicationSnapshotController(IMediator mediator)
+        {
+            _mediator = mediator;
+        }
+
+        /// <summary>
+        ///     Creates a snapshot of the requested application. Used for RoATP
```
Comment (davidgouge) [inline src/SFA.DAS.QnA.Api/Controllers/ApplicationSnapshotController.cs]: Ugh, client specific comment. 😉

Comment (martin-hornsby) [reply src/SFA.DAS.QnA.Api/Controllers/ApplicationSnapshotController.cs]: yeah it's a big stinky stain within QnA 😿 I've added some remarks to further expand on why the endpoint is there

File: src/SFA.DAS.QnA.Application.UnitTests/CommandsTests/CreateSnapshotTests/When_application_exists.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+using NUnit.Framework;
+using SFA.DAS.QnA.Application.Commands.CreateSnapshot;
+
+namespace SFA.DAS.QnA.Application.UnitTests.CommandsTests.CreateSnapshotTests
+{
+    public class When_application_exists : CreateSnapshotTestBase
+    {
+        [Test]
+#if (!DEBUG)
+        [Ignore("Must be ran on local DEV machine")]
```
Comment (MarkFCain) [inline src/SFA.DAS.QnA.Application.UnitTests/CommandsTests/CreateSnapshotTests/When_application_exists.cs]: run not run!

File: src/SFA.DAS.QnA.Application.UnitTests/CommandsTests/CreateSnapshotTests/When_application_not_found.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+using NUnit.Framework;
+using SFA.DAS.QnA.Application.Commands.CreateSnapshot;
+
+namespace SFA.DAS.QnA.Application.UnitTests.CommandsTests.CreateSnapshotTests
+{
+    public class When_application_not_found : CreateSnapshotTestBase
+    {
+        [Test]
+#if (!DEBUG)
+        [Ignore("Must be ran on local DEV machine")]
```
Comment (MarkFCain) [inline src/SFA.DAS.QnA.Application.UnitTests/CommandsTests/CreateSnapshotTests/When_application_not_found.cs]: RUN!

File: src/SFA.DAS.QnA.Application.UnitTests/CommandsTests/CreateSnapshotTests/When_application_exists.cs (hunk 6 above)
Comment (martin-hornsby) [reply src/SFA.DAS.QnA.Application.UnitTests/CommandsTests/CreateSnapshotTests/When_application_exists.cs]: sorted

File: src/SFA.DAS.QnA.Application.UnitTests/CommandsTests/CreateSnapshotTests/When_application_not_found.cs (hunk 7 above)
Comment (martin-hornsby) [reply src/SFA.DAS.QnA.Application.UnitTests/CommandsTests/CreateSnapshotTests/When_application_not_found.cs]: sorted

**Comment by alandburns** [comment]
Is new code for a new feature, and should not affect existing consumers of the QnA API.

**Pull request merged**